### ISSUE TITLE
Various illegal/bad instruction warnings 

### ISSUE THREAD

**Issue by SolidStill** [dosdude1/macos-catalina-patcher#26]
Please only file an issue here for code-related problems. This is not an issue system for resolving issues with using the compiled binaries.
- [x ] I understand

Firstly, thank-you for doing God's work, can't praise you enough! 

So, I'm getting some heavy duty warnings when trying to `brew install fzf`/`brew install ruby` - the guy at the Homebrew Discourse said "Your system just seems very broken. This is beyond anything I’ve ever seen and I’d say the different languages failing with native stacktraces are an indication that it’s going wrong on an OS level. I’d make backups of everything on the device and reinstall macOS."

The warnings, crash reports etc. are linked below...
Please let me know if you'd prefer the relevant info to be posted here, np.

Homebrew discourse & Github issue links: 
- RE fzf: https://discourse.brew.sh/t/fzf-corrupted-install-via-bottle-and-also-via-source-build/7236?u=solidstill & https://github.com/junegunn/fzf/issues/1935#issue-585535541
- RE ruby: https://discourse.brew.sh/t/post-install-illegal-instruction-warning-installing-ruby-2-7-0-catalina-bottle-1-tar-gz/7246

Thanks a million

**Comment by dosdude1** [comment]
There's a flag you have to set, I forgot what exactly it is. Essentially, it prevents brew from using modern instruction sets, like SSE4.2.

**Comment by SolidStill** [comment]
Right, I see... I tested the above on Mojave and threw the same errors - thought I was going crazy. Thank you for the heads-up @dosdude1 .

I'm looking for the relevant info RE processor instruction sets right now, and I'm operating way above my pay grade here, but I'll summarise what I'm seeing and thinking as best I can in case someone else finds this useful: 

- I'm on a MacBookPro5,2 (A1297), having a Core 2 Duo processor running at 2.8 GHz with 6MB of L2 Cache. 

- Looking at the [list of Core 2 Duos on wikipedia](https://en.wikipedia.org/wiki/List_of_Intel_Core_2_microprocessors#Core_2_Duo) with those specs, as far as I can tell this is a chip with Penryn microarchitecture. 

- The [Penryn Wikipedia page](https://en.wikipedia.org/wiki/Penryn_(microarchitecture)) lists various instruction sets, mentioning various SSE iterations going as high as SSE4.1 (and, critically, not as high as SSE4.2, as you have suggested).

- When I searched for info RE Homebrew and compilation-flags/instructions-sets/SSE I'm led to a github discussion amongst members of the Homebrew team debating pros/cons of supporting particular generations of processors here: https://github.com/Homebrew/brew/issues/5425. **The key flag mentioned is `-march=core2`** (ie 'micro-architecture=Core2Duo') as opposed to the current default for Mojave of `-march=nehalem` which permits SSE4.2 instructions. 

- I was not able to pass this `-march=core2` flag as an argument of a `$ brew install...` type command. But the discussion contains a comment mentioning a suitable approach using what I think is a single-line build-from-source command with the syntax `HOMEBREW_OPTFLAGS="-march=core2" brew install -s ruby`, but this gives me errors too:

```
➜  ~ HOMEBREW_OPTFLAGS="-march=core2" brew install -s ruby
Updating Homebrew...
Successfully rebased and updated refs/heads/master.
==> Auto-updated Homebrew!
Updated 1 tap (homebrew/cask).
No changes to formulae.

==> Installing dependencies for ruby: pkg-config
==> Installing ruby dependency: pkg-config
==> Downloading https://homebrew.bintray.com/bottles/pkg-config-0.29.2.mojave.bo
==> Downloading from https://akamai.bintray.com/dc/dcfcba723621eee085ddd5ebadc41
######################################################################## 100.0%
==> Pouring pkg-config-0.29.2.mojave.bottle.1.tar.gz
🍺  /usr/local/Cellar/pkg-config/0.29.2: 11 files, 626.9KB
==> Installing ruby
==> Downloading https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.0.tar.xz
######################################################################## 100.0%
==> ./configure --prefix=/usr/local/Cellar/ruby/2.7.0 --enable-shared --with-sit
==> make
Last 15 lines from /Users/chris/Library/Logs/Homebrew/ruby/02.make:
7fffa8e00000-7fffa9000000 rw- /usr/lib/dyld
7fffa9000000-7fffa9200000 rw- /usr/lib/dyld
7fffa9200000-7fffa9387000 rw- /usr/lib/dyld
7fffa9387000-7fffc0000000 r-- /usr/lib/dyld
7fffc0000000-7fffffe00000 r-- /usr/lib/dyld
7fffffe00000-7fffffe01000 r-- /usr/lib/dyld
7ffffffdd000-7ffffffde000 r-x /usr/lib/dyld
[IMPORTANT]
Don't forget to include the Crash Report log file under
DiagnosticReports directory in bug reports.

make: *** [encdb.h] Abort trap: 6
make: *** Waiting for unfinished jobs....
./tool/ifchange "--timestamp=.rbconfig.time" rbconfig.rb rbconfig.tmp
rbconfig.rb updated

READ THIS: https://docs.brew.sh/Troubleshooting
```
(The above started by downloading a binary (as opposed to compiling a core2 build) of pkgconfig, is that maybe where this particular installation has gone wrong? Surely not all binaries are suspect though, since this Mojave installation and the software running on it functions seemingly flawlessly?) 

The gist I got from reading around this is that Homebrew is very much not into providing "custom build optimisation" flags. My knowledge is very much limited and shallow but I feel like I'm close to, if indeed there is, a reasonable solution wrt to setting specific flags; the holy grail here being a method of setting this as a default via a shell variable or similar, I'd dare guess.

Any guidance, or even just vague pointing in the right direction, would be awesome. Thanks!

**Comment by SolidStill** [comment]
Initially, I wondered if I could set my brew config to be detected as High Sierra since the HighSierra packages are still compiled for "core2" compatibility AFAICT; but not being able to see how to do this, or what the undesired side-effects might be, I haven't pursued this approach.

After some more digging:

- Eventually, I came across this discussion: https://discourse.brew.sh/t/almost-all-applications-crash-with-illegal-hardware-instruction-on-old-2009-macbook-pro/4089/5 - with the instruction to "Change `if version >= :mojave` to something like `if version >= :mojave && false`" in the file '/usr/local/Homebrew/Library/Homebrew/extend/os/mac/hardware.rb', "And then additionally you’ll need to manually --build-from-source for everything you install". I googled around for "/usr/local/Homebrew/Library/Homebrew/extend/os/mac/hardware.rb core2", and found someone on MacRumors (https://forums.macrumors.com/threads/macos-10-14-mojave-on-unsupported-macs-thread.2121473/page-466?post=27026470#post-27026470) who swapped the `:nehalem` line for `:core2` (using the following command: `sed -i -e 's/:nehalem/:core2/g' /usr/local/Homebrew/Library/Homebrew/extend/os/mac/hardware.rb`).

- But I've got a lot of existing brew installs on my system; and even though I think the vast majority are OK (no warnings AFAIK), I don't know for sure. So I then found "To rebuild all installed FORMULAE from source": `brew list | xargs brew reinstall --build-from-source`. This command reinstalled all existing packages by building from source except for a few previously-uninstalled dependencies which it poured when trying to build libyaml:
```
==> Pouring automake-1.16.2.mojave.bottle.tar.gz
==> Pouring libtool-2.4.6_1.mojave.bottle.tar.gz
==> Pouring autoconf-2.69.mojave.bottle.4.tar.gz
```
Maybe homebrew used "nehalem"-built bottles for dependencies that had already been installed? I just don't know if it did, and I don't know if that's even a problem if I'm not seeing any warnings from the compilation of libyaml.
Maybe homebrew has automatically selected for binaries that are "core2" compatible now that I've changed the CPU flag in hardware.rb - but I doubt it given the following explicit instruction from the discourse.brew.sh post mentioned earlier: "And then additionally you’ll need to manually --build-from-source for everything you install", which seems to imply any *.mojave.bottle binary needs to substitued with a source build.

- When looking to make --build-from-source the default method for all Homebrew installs there used to be a "HOMEBREW_BUILD_FROM_SOURCE variable", but it "was deprecated and removed". So, I'm still on the hunt for a method making all `brew ...` commands compatible with my system, without modificaiton, by default.
